src/System.Text.Json/src/System/Text/Json/Serialization/ReflectionEmitMemberAccessor.cs

```diff
@@ -0,0 +1,150 @@
+                typeof(ReflectionEmitMemberAccessor).Module,
+                skipVisibility: true);
+
+            ILGenerator generator = dynamicMethod.GetILGenerator();
+
+            if (realMethod == null)
+            {
+                LocalBuilder local = generator.DeclareLocal(type);
+
+                generator.Emit(OpCodes.Ldloca_S, local);
+                generator.Emit(OpCodes.Initobj, type);
+                generator.Emit(OpCodes.Ldloc, local);
+                generator.Emit(OpCodes.Box, type);
+            }
+            else
+            {
+                generator.Emit(OpCodes.Newobj, realMethod);
+            }
+
+            generator.Emit(OpCodes.Ret);
+
+            return (JsonClassInfo.ConstructorDelegate)dynamicMethod.CreateDelegate(typeof(JsonClassInfo.ConstructorDelegate));
+        }
+
+        public override object ImmutableCollectionCreateRange(Type constructingType, Type elementType)
+        {
+            MethodInfo createRange = ImmutableCollectionCreateRangeMethod(constructingType, elementType);
+
+            if (createRange == null)
+            {
+                return null;
+            }
+
+            return createRange.CreateDelegate(
+                typeof(DefaultImmutableConverter.ImmutableCreateRangeDelegate<>).MakeGenericType(elementType), null);
+        }
+
+        public override object ImmutableDictionaryCreateRange(Type constructingType, Type elementType)
+        {
+            MethodInfo createRange = ImmutableDictionaryCreateRangeMethod(constructingType, elementType);
+
+            if (createRange == null)
+            {
+                return null;
+            }
+
+            return createRange.CreateDelegate(
+                typeof(DefaultImmutableConverter.ImmutableDictCreateRangeDelegate<,>).MakeGenericType(typeof(string), elementType), null);
+        }
+
+        public override Func<object, TProperty> CreatePropertyGetter<TClass, TProperty>(PropertyInfo propertyInfo) =>
+            (Func<object, TProperty>)CreatePropertyGetter(propertyInfo, typeof(TClass));
+
+        private static Delegate CreatePropertyGetter(PropertyInfo propertyInfo, Type classType)
+        {
+            MethodInfo realMethod = propertyInfo.GetGetMethod();
+            Type objectType = typeof(object);
+
+            var dynamicMethod = new DynamicMethod(
+                realMethod.Name,
+                propertyInfo.PropertyType,
+                new[] { objectType },
+                typeof(ReflectionEmitMemberAccessor).Module,
+                skipVisibility: true);
+
+            ILGenerator generator = dynamicMethod.GetILGenerator();
+
+            generator.Emit(OpCodes.Ldarg_0);
+
+            if (classType.IsValueType)
+            {
+                generator.Emit(OpCodes.Unbox_Any, classType);
+                generator.Emit(OpCodes.Call, realMethod);
+            }
+            else
+            {
+                generator.Emit(OpCodes.Castclass, classType);
```

Comment:
Is `Castclass` better\faster than `Ldarg_0`?. Also see similar comment below for setter.

Reply:
Moreover the perf could be improved by removing `castclass` but at the cost of type safety. I've posted the requested benchmark results above.